### PR TITLE
Fix dmypy run on Windows

### LINKED ISSUE #10709: `dmypy run` won't run, prints `"Restarting: configuration changed"` when `ignore_missing_imports = True` is configured
**Bug Report**

Since at least `v0.902`, executing `dmypy run` in a project that has `ignore_missing_imports = True` in `mypy.ini` is not working anymore.

**To Reproduce**

(Write your steps here:)

1. Create `src/something.py` with some code in it, and create `mypy.ini` with:
```ini
[mypy]
files = src
ignore_missing_imports = True
```
2. Run `dmypy run`

**Expected Behavior**

`dmypy run` should execute correctly, outputting something like:

```
Daemon started
src\something.py:1: error: Incompatible types in assignment (expression has type "str", variable has type "int")
Found 1 error in 1 file (checked 26 source files)
```

**Actual Behavior**

`dmypy run` keeps showing:

```
Restarting: configuration changed
Daemon stopped
Daemon started
Response: {'restart': 'configuration changed', 'platform': 'win32', 'python_version': '3_9', 'roundtrip_time': 0.5543885231018066}
```

even when trying to run it again.

**Your Environment**

- Mypy version used: `v0.902`, `v0.910` and `0.920+dev.379622de482ec16dd1676ad6d183dcfca5203194`
- Mypy command-line flags: `dmypy run`
- Mypy configuration options from `mypy.ini` (and other config files):
```
[mypy]
files = src
ignore_missing_imports = True
```
- Python version used: 3.9.5
- Operating system and version: Windows 10 21H1

### PR DESCRIPTION
**Pull request by ilevkivskyi**
Fixes #10709 

Two changes here:
* Add equality methods to `ErrorCode` because they may appear in options snapshots
* Use stable comparison for option snapshots, since `Options.apply_changes()` does some unrelated things with flags

Both are only relevant on Windows where options may roundtrip as: `options -> snapshot -> pickle -> base64 -> unpickle -> apply_changes`.

### CODE REVIEW

**Comment by Avasam** [comment]
I `pip install git+https://github.com/ilevkivskyi/mypy.git@fix-dmypy-run#egg=mypy` into my typeshed environment, re-enabled `--ignore-missing-imports`, and it seems to be working correctly now.

The image below shows that missing imports are successfully ignored, and that mypy does show errors where it should
image

**Comment by ilevkivskyi** [comment]
OK, great, thanks for confirming!

**Comment by Avasam** [comment]
I'd like if someone else can confirm too, since my usage is going through a 3rd party extension. Jsut to make sure it wasn't a fluke.

**Comment by AlexWaygood** [comment]
I can also repro the issue on a Windows machine, and can also confirm that this patch appears to fix it!

**Review by svalentin — APPROVED** [review]
Tested it and seems to work. 

before:
```
PS C:\src\mypy-play> python -m mypy.dmypy run .\test.py
Restarting: configuration changed
Daemon stopped
Daemon started
Response: {'restart': 'configuration changed', 'stdout': '', 'stderr': '', 'platform': 'win32', 'python_version': '3_11', 'roundtrip_time': 0.488344669342041}
```

after applying PR:
```
PS C:\src\mypy-play> python -m mypy.dmypy run .\test.py
Restarting: configuration changed
Daemon stopped
Daemon started
←[1m←[92mSuccess: no issues found in 42 source files←[0m
PS C:\src\mypy-play> python -m mypy.dmypy run .\test.py
←[1m←[92mSuccess: no issues found in 1 source file←[0m
PS C:\src\mypy-play>
```

File: mypy/dmypy_server.py [hunk 1]
```diff
@@ -170,11 +170,17 @@ class Server:
     # NOTE: the instance is constructed in the parent process but
     # serve() is called in the grandchild (by daemonize()).
 
-    def __init__(self, options: Options, status_file: str, timeout: int | None = None) -> None:
+    def __init__(
+        self,
+        options: Options,
+        status_file: str,
+        timeout: int | None = None,
+        options_snapshot: dict[str, object] | None = None,
```
Comment (svalentin) [inline mypy/dmypy_server.py:178]: It feels like it shouldn't be necessary to pass both options and options_snapshot. One or the other should be enough.

Comment (ilevkivskyi) [reply mypy/dmypy_server.py:178]: Yeah, but then we will need something like "stable compare", which will be a tiny bit slower (this is btw the original way I did it). OTOH, then motivation is much more obvious, so let's go that way indeed.

Comment (ilevkivskyi) [reply mypy/dmypy_server.py]: @svalentin Could you please test again just in case?

**Comment by ilevkivskyi** [comment]
@hauntsaninja It looks like `mypy_primer` is broken because `AutoSplit` fails to install. I tried re-running but it didn't help: https://github.com/python/mypy/actions/runs/5265892753/jobs/9519524776?pr=15429 Could you please take a look?

**Comment by AlexWaygood** [comment]
> @hauntsaninja It looks like `mypy_primer` is broken because `AutoSplit` fails to install. I tried re-running but it didn't help: https://github.com/python/mypy/actions/runs/5265892753/jobs/9519524776?pr=15429 Could you please take a look?

Looks like the issue is that winsdk can't be installed (maybe it's a Windows-only package?). If so then hauntsaninja/mypy_primer#89 should fix it.

winsdk was only added to the mypy-primer recipe for Autosplit yesterday, in https://github.com/hauntsaninja/mypy_primer/pull/87

File: mypy/dmypy_server.py (hunk 1 above)
Comment (svalentin) [reply mypy/dmypy_server.py]: Nice!
Still works:

```
PS C:\src\mypy-play> python -m mypy.dmypy run .\test.py
Success: no issues found in 1 source file
```

Comment (svalentin) [reply mypy/dmypy_server.py]: For what it's worth, I think this is a much more elegant fix

**Review by svalentin — APPROVED** [review]

**Pull request merged**